### ISSUE TITLE
Add ProtocolBlockTypePalette

### ISSUE THREAD

**Issue by madmaxoft** [cuberite/cuberite#4375]
We need a class that will use `BlockState`, together with data loaded from the vanilla Generators (https://wiki.vg/Data_Generators), to provide the mapping between blocks with state to numbers used by the specific protocol. This issue is about creating the class (not plugging it into a protocol just yet).

Basically, we need a class that provides the following APIs:
```cpp
UInt32 index(const AString & aBlockTypeName, const BlockState & aBlockState);
bool loadFromString(const AString & aEntireMapping);
```
The `index` function will return the palette index for the specified block, or a special number indicating that the block is not present in the mapping.

The `loadFromString` function will load the entire mapping from a string and return `true` on success, `false` (and log exact reason to console) on failure. The input string should be a reasonable simplification of the report generated by the vanilla server, I propose the following format:
```
ProtocolBlockTypePalette
0	minecraft:air
...
8	minecraft:grass_block	snowy	true
9	minecraft:grass_block	snowy	false
...
144	minecraft:oak_leaves	distance	1	persistent	true
...
```
Basically the first line identifies the file kind, then each line has tab separated values of the protocol ID, block type name and then all the BlockState properties.
A pure Lua script to turn the Generator report into this format should be a part of the implementation, too.
The `ProtocolBlockTypePalette` class should have its own tests (for inspiration look at the BlockState tests in /tests/BlockTypeRegistry/BlockStateTest.cpp)
Since the parsing will only be done on server startup, it doesn't need to be super-efficient, simple string splitting should be sufficient; there are lots of helper functions in StringUtils.

**Comment by madmaxoft** [comment]
Alternatively, the `loadFromString` function could report errors by throwing a specialized exception. This could be more useful once we chain to that function to make it load from files, multiple files etc. Ideally the exception should have an error message plus the exact place in the string (line number) where the error occurred.

**Comment by Aplaus228** [comment]
I cant understand how need parse server if the data generator available from Mojang? but this is another project. Also, I haven't seen protocol id for blocks its must be found at the file or we must give a unique id for every block?

**Comment by Seadragon91** [comment]
The protocol ids for items, block and .... are in file `registries.json`.

**Comment by Aplaus228** [comment]
I would try to do it. if someone has started this work please write a comment so that  I don't do double work.

**Comment by Aplaus228** [comment]
sry but I am stuck in the beginning I can't understand how I can parse JSON file with Lua.
someone can direct me to the correct way?

**Comment by madmaxoft** [comment]
Any progress on this? We will be needing it soon.

**Comment by E14** [comment]
@madmaxoft On it, the hardest part for me is done. I tried to keep everything as tidy as possible, however I need a change in the `BlockState` class as the registry needs to be able to access the checksum (at least). Currently I'm using an accessor to `mChecksum` called `checksum` - should I make a separate PR for that or put it in one?
(May also need to improve the checksum to be more unique though)

About the implementation, I'm assuming that `index(...)` will be called fairly frequently and that it's not good to iterate over the whole palette all the time, so I'm using a hashmap.

**Comment by madmaxoft** [comment]
What do you need the checksum for? It shouldn't be accessible to the outside. It's not even guaranteed to be unique. It can only be used for the fast equality check, as implemented in the `==` operator.

**Comment by E14** [comment]
Either need the checksum, an actual hash, access to the key list, or access to mState, since otherwise there is no way index by block state

**Comment by madmaxoft** [comment]
Wouldn't a `BlockState::operator <` be enough for indexing by block state? That way, the checksum would still be contained within. The operator could be simply to compare the checksums, and if equal, then somehow compare the `mState`.

**Comment by E14** [comment]
Hmm, maybe I lost myself there - it should work with a 2-d hashmap/map structure, since there are only a few states per blocktype.
